c_src/main.c

```diff
@@ -281,6 +281,89 @@ static inline ERL_NIF_TERM erlfdb_future_get_keyvalue_array(ErlNifEnv *env,
     }
 }
 
+static inline ERL_NIF_TERM
+erlfdb_future_get_mappedkeyvalue_array(ErlNifEnv *env, ErlFDBFuture *f) {
+    FDBMappedKeyValue const *mappedkvs;
+    FDBKeyValue *data;
+    fdb_bool_t more;
+    int count;
+    unsigned char *buf;
+    ERL_NIF_TERM pq_key;
+    ERL_NIF_TERM pq_val;
+    ERL_NIF_TERM sq_begin;
+    ERL_NIF_TERM sq_end;
+    ERL_NIF_TERM key;
+    ERL_NIF_TERM val;
+    ERL_NIF_TERM ret;
+    ERL_NIF_TERM kvs_ret;
+    fdb_error_t err;
+    int i, j;
+
+    err = fdb_future_get_mappedkeyvalue_array(f->future, &mappedkvs, &count,
+                                              &more);
+    if (err != 0) {
+        return erlfdb_erlang_error(env, err);
+    }
+
+    ret = enif_make_list(env, 0);
+
+    for (i = count - 1; i >= 0; i--) {
+        // pq_key
+        buf = enif_make_new_binary(env, mappedkvs[i].key.key_length, &pq_key);
+        memcpy(buf, mappedkvs[i].key.key, mappedkvs[i].key.key_length);
+
+        // pq_val (which is also a key)
+        buf = enif_make_new_binary(env, mappedkvs[i].value.key_length, &pq_val);
+        memcpy(buf, mappedkvs[i].value.key, mappedkvs[i].value.key_length);
+
+        const FDBGetRangeReqAndResult *getRange = &mappedkvs[i].getRange;
+
+        // sq_begin
+        buf = enif_make_new_binary(env, getRange->begin.key.key_length,
+                                   &sq_begin);
+        memcpy(buf, getRange->begin.key.key, getRange->begin.key.key_length);
+
+        // sq_end
+        buf = enif_make_new_binary(env, getRange->end.key.key_length, &sq_end);
+        memcpy(buf, getRange->end.key.key, getRange->end.key.key_length);
+
+        data = getRange->data;
+
+        kvs_ret = enif_make_list(env, 0);
+        for (j = getRange->m_size - 1; j >= 0; j--) {
+
+            // mapped key
+            buf = enif_make_new_binary(env, data[j].key_length, &key);
+            memcpy(buf, data[j].key, data[j].key_length);
+
+            // mapped val
+            buf = enif_make_new_binary(env, data[j].value_length, &val);
+            memcpy(buf, data[j].value, data[j].value_length);
+
+            kvs_ret = enif_make_list_cell(env, T2(env, key, val), kvs_ret);
+        }
+
+        // {
+        //   {PrimaryQueryKey, PrimaryQueryValue},
+        //   {SecondaryQueryKeyBegin, SecondaryQueryKeyEnd},
+        //   [
+        //      {SecondaryQueryKey, SecondaryQueryValue},
+        //      ...
+        //   ]
+        // }
+        ret = enif_make_list_cell(env,
+                                  T3(env, T2(env, pq_key, pq_val),
+                                     T2(env, sq_begin, sq_end), kvs_ret),
+                                  ret);
+    }
+
+    if (more) {
+        return T3(env, ret, enif_make_int(env, count), ATOM_true);
```

Comment:
The data structures returned here for more=true will cause errors in `erlfdb:fold_range_int/4`

Excerpt:
```erlang
    {RawRows, Count, HasMore} = wait(Future),

    Count = length(RawRows),
...
            LastKey = element(1, lists:last(Rows))
```